### PR TITLE
Improve and Expand API Documentation

### PR DESCRIPTION
**Pull request by omad**
### Reason for this pull request

The existing API documentation is both really hard to browse, and is missing lots and lots of valuable functionality.

### Proposed changes
- Flatten out the structure of the current API docs so that everything is listed on a single `Ctrl-F` able page.
- Add lots of currently unmentioned functions/methods and classes into the front API docs page.

See [proposed new API page](https://damien-agdc.readthedocs.io/en/latest/dev/api/index.html) for a view of the new page. Compare with [current API page](https://datacube-core.readthedocs.io/en/latest/dev/api/index.html) 

 - ~[ ] Closes #xxxx~
 - ~[ ] Tests added / passed~
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

**Comment by mpaget** [comment]
Looks good, @omad. Nice initiative, makes browsing the scope and discovering specific functions much easier.

Direct links to each function is helpful for more information. A spot check suggests this is done for most functions but is missing (if applicable) for:
- Product Querying
- Product Addition/Modification
And it might be helpful to have corresponding text descriptions for the "Creating Geometries" functions.

Also, it would be good to do something similar (flatten and incorporate into this page) for the "Everything Else"s but they can be a separate pull request.

File: docs/dev/api/index.rst [hunk 1]
```diff
@@ -1,18 +1,351 @@
...
+   get_locations
+   get_archived_locations
+   get_datasets_for_location
+
+
+Dataset Writing
+---------------
+
+When connected to an ODC Database, these methods are available for
+adding, updating and deleting datasets::
```
Comment (Kirill888) [inline docs/dev/api/index.rst]: except we actually can not **delete** datasets, only **archive** datasets.

File: docs/dev/api/index.rst [hunk 2]
```diff
@@ -1,18 +1,351 @@
...
+Open Data Cube includes a set of CRS aware geometry utilities.
+
+Geometry Classes
+----------------
+
+.. autosummary::
+   :nosignatures:
+   :toctree: generate/
+
+   utils.geometry.BoundingBox
```
Comment (Kirill888) [inline docs/dev/api/index.rst:259]: `BoundingBox` is not "CRS aware", and it's really annoying that it is the odd one out.

**Review by Kirill888 — DISMISSED** [review]
Looks good, see some minor notes below. Also there are still a lot of places where we refer to "products" as "dataset types", haven't seen them in the diff but there are a few in the "Product Querying" section, I guess they get pulled from python doc strings.

**Review by andrewdhicks — DISMISSED** [review]
:+1:

**Review by Kirill888 — APPROVED** [review]

**Pull request merged**